File: app/src/main/java/com/twilio/video/app/util/StatsScheduler.java
/*
 * Copyright (C) 2017 Twilio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twilio.video.app.util;

import android.os.Handler;
import android.os.HandlerThread;
import android.support.annotation.NonNull;
import com.twilio.video.Room;
import com.twilio.video.StatsListener;

public class StatsScheduler {
  private HandlerThread handlerThread;
  private Handler handler;

  public StatsScheduler() {}

  // Listener will be called from scheduler thread
  public void scheduleStatsGathering(
      final @NonNull Room room,
      final @NonNull StatsListener listener,
      final long delayInMilliseconds) {
    if (room == null) {
      throw new NullPointerException("Room must not be null");
    }
    if (listener == null) {
      throw new NullPointerException("StatsListener must not be null");
    }
    if (isRunning()) {
      cancelStatsGathering();
    }
    this.handlerThread = new HandlerThread("StatsSchedulerThread");
    handlerThread.start();
    handler = new Handler(handlerThread.getLooper());

    final Runnable statsRunner =
        new Runnable() {
          @Override
          public void run() {
            room.getStats(listener);
            handler.postDelayed(this, delayInMilliseconds);
          }
        };
    handler.post(statsRunner);
  }

  public boolean isRunning() {
    return (handlerThread != null && handlerThread.isAlive());
  }

  public void cancelStatsGathering() {
    if (handlerThread != null && handlerThread.isAlive()) {
      handlerThread.quit();
      handlerThread = null;
    }
    handler = null;
  }
}

File: app/src/main/java/com/twilio/video/app/sdk/StatsScheduler.kt
/*
 * Copyright (C) 2019 Twilio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.twilio.video.app.sdk

import android.os.Handler
import android.os.HandlerThread
import com.twilio.video.Room
import com.twilio.video.StatsListener
import timber.log.Timber

class StatsScheduler(private val roomManager: RoomManager, private val room: Room) {
    private var handlerThread: HandlerThread? = null
    private var handler: Handler? = null
    private val statsListener: StatsListener = StatsListener { statsReports ->
        roomManager.sendStatsUpdate(statsReports)
    }
    private val isRunning: Boolean
        get() = handlerThread?.isAlive ?: false

    fun start() {
        if (isRunning) {
            stop()
        }
        val handlerThread = HandlerThread("StatsSchedulerThread")
        this.handlerThread = handlerThread
        handlerThread.start()
        val handler = Handler(handlerThread.looper)
        this.handler = handler
        val statsRunner: Runnable = object : Runnable {
            override fun run() {
                room.getStats(statsListener)
                handler.postDelayed(this, 1000)
            }
        }
        handler.post(statsRunner)
        Timber.d("Stats scheduler thread started")
    }

    fun stop() {
        if (isRunning) {
            handlerThread?.let { handlerThread ->
                handlerThread.quit()
                this.handlerThread = null
                handler = null
                Timber.d("Stats scheduler thread closed")
            }
        }
    }
}
